ctlra/devices/midi_generic.c

```diff
@@ -0,0 +1,199 @@
+};
+
+static uint32_t
+midi_generic_poll(struct ctlra_dev_t *base)
+{
+	struct midi_generic_t *dev = (struct midi_generic_t *)base;
+	ctlra_midi_input_poll(dev->midi);
+	return 0;
+}
+
+int
+midi_generic_midi_input_cb(uint8_t nbytes, uint8_t * buf, void *ud)
+{
+	struct midi_generic_t *dev = (struct midi_generic_t *)ud;
+
+	switch(buf[0] & 0xf0) {
+	case 0x90: /* Note On */
+	case 0x80: /* Note Off */ {
+		struct ctlra_event_t event = {
+			.type = CTLRA_EVENT_BUTTON,
+			.button  = {
+				.id = buf[1],
+				.pressed = buf[0] >= 0x90,
+				.has_pressure = 1,
+				.pressure = buf[2] / 127.f,
+			},
+		};
+		struct ctlra_event_t *e = {&event};
+		dev->base.event_func(&dev->base, 1, &e,
+				     dev->base.event_func_userdata);
+		} break;
+
+	case 0xb0: /* control change */ {
+		struct ctlra_event_t event = {
+			.type = CTLRA_EVENT_SLIDER,
+			.slider  = {
+				.id = buf[1],
+				.value = buf[2] / 127.f
+			},
+		};
+		struct ctlra_event_t *e = {&event};
+		dev->base.event_func(&dev->base, 1, &e,
+				     dev->base.event_func_userdata);
+		}
+		break;
+	};
+
+	return 0;
+}
+
+static const char *
+midi_generic_control_get_name(enum ctlra_event_type_t type,
+			      uint32_t control)
+{
+	return "generic";
+}
+
+static void midi_generic_light_set(struct ctlra_dev_t *base,
+				    uint32_t light_id,
+				    uint32_t light_status)
+{
+	struct midi_generic_t *dev = (struct midi_generic_t *)base;
+
+	uint8_t out[3];
+
+	if(light_id == -1) {
+		out[2] = (light_status >> 16) & 0xff;
+		out[1] = (light_status >>  8) & 0xff;
+		out[0] = (light_status >>  0) & 0xff;
+	} else {
+		uint8_t b3 = 0;
+		b3 |= light_status >> 24;
+		b3 |= light_status >> 16;
+		b3 |= light_status >>  8;
+
+		out[0] = 0x90;
+		out[1] = light_id;
+		out[2] = b3;
+	}
+
+	ctlra_midi_output_write(dev->midi, 3, out);
+}
+
+void
+midi_generic_light_flush(struct ctlra_dev_t *base, uint32_t force)
+{
+	struct midi_generic_t *dev = (struct midi_generic_t *)base;
+}
+
+static int32_t
+midi_generic_disconnect(struct ctlra_dev_t *base)
+{
+	struct midi_generic_t *dev = (struct midi_generic_t *)base;
+	free(dev);
+	return 0;
+}
+
+struct ctlra_dev_info_t ctlra_midi_generic_info;
+
+struct ctlra_dev_t *
+ctlra_midi_generic_connect(ctlra_event_func event_func, void *userdata,
+			   void *future)
+{
+	(void)future;
+	struct midi_generic_t *dev = calloc(1, sizeof(struct midi_generic_t));
+	if(!dev)
+		goto fail;
+
+	dev->midi = ctlra_midi_open("Ctlra Generic",
+				    midi_generic_midi_input_cb, dev);
+	if(dev->midi == 0) {
+		printf("Ctlra: error opening midi i/o\n");
+		goto fail;
+	}
+
+	dev->base.info = ctlra_midi_generic_info;
+
+	dev->base.poll = midi_generic_poll;
+	dev->base.disconnect = midi_generic_disconnect;
+	dev->base.light_set = midi_generic_light_set;
+	dev->base.light_flush = midi_generic_light_flush;
+
+	dev->base.event_func = event_func;
+	dev->base.event_func_userdata = userdata;
+
+	return (struct ctlra_dev_t *)dev;
+fail:
+	free(dev);
+	return 0;
```

Comment:
I'd be tempted to put a NULL here just to highlight that this function is pointery (wheras the others in the file are inty).